src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -389,6 +466,206 @@ ForeignConstraintFindDistKeys(HeapTuple pgConstraintTuple,
 }
 
 
+/*
+ * ErrorIfUnsupportedAlterAddDropFKeyBetweenReferecenceAndLocalTable runs checks related to ALTER
+ * ADD / DROP foreign key constraints between local tables and reference tables.
+ *
+ * If constraint is not NULL, then we perform checks for ADD command, otherwise we check for DROP.
+ */
+void
+ErrorIfUnsupportedAlterAddDropFKeyBetweenReferecenceAndLocalTable(Oid referencingTableOid,
+																  Oid
+																  referencedTableOid,
+																  Constraint *constraint)
+{
+	bool referencingIsDistributed = IsDistributedTable(referencingTableOid);
+	char referencingDistMethod = 0;
+	bool referencingIsReferenceTable = false;
+
+	bool referencedIsDistributed = IsDistributedTable(referencedTableOid);
+	char referencedDistMethod = 0;
+	bool referencedIsReferenceTable = false;
+
+	if (referencingIsDistributed)
+	{
+		referencingDistMethod = PartitionMethod(referencingTableOid);
+		referencingIsReferenceTable = (referencingDistMethod == DISTRIBUTE_BY_NONE);
+	}
+
+	if (referencedIsDistributed)
+	{
+		referencedDistMethod = PartitionMethod(referencedTableOid);
+		referencedIsReferenceTable = (referencedDistMethod == DISTRIBUTE_BY_NONE);
+	}
+
+	/* reference table <> local table */
+	if ((referencingIsReferenceTable && !referencedIsDistributed) ||
+		(!referencingIsDistributed && referencedIsReferenceTable))
+	{
+		/* constraint is not NULL when the command is ALTER TABLE ADD constraint */
+		if (constraint != NULL)
+		{
+			/*
+			 * "ALTER TABLE reference_table ADD CONSTRAINT fkey FOREIGN KEY REFERENCES
+			 * local_table ON DELETE (UPDATE) CASCADE" commands are not supported
+			 */
+			bool onDeleteOrOnUpdateCascade = (constraint->fk_upd_action ==
```

Comment:
`CASCADE` is not the only option we will have problem with. We should error out on everything "except" `FKCONSTR_ACTION_NOACTION` and `FKCONSTR_ACTION_RESTRICT`. See https://www.postgresql.org/docs/current/sql-createtable.html for what each of these mean (the options weren't listed in ALTER TABLE's doc page, so I linked to the CREATE TABLE page)

Reply:
thanks, I will apply needed changes